Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
 }
 
 /**
  @Status Interoperable
 */
-- (id)init {
-    if (self = [super init]) {
-        priv = new NSOperationQueuePriv();
-
-        priv->_maxConcurrentOperationCount = 1;
-        priv->workAvailable = [[NSCondition alloc] init];
-        priv->suspendedCondition = [[NSCondition alloc] init];
-        priv->allWorkDone = [[NSCondition alloc] init];
-        priv->isSuspended = 0;
-
-        _completionQueue = dispatch_queue_create("NSOperation finish queue", nullptr);
-    }
-
-    return self;
-}
-
-- (id)_initMainThread {
-    priv = new NSOperationQueuePriv();
-
-    priv->_maxConcurrentOperationCount = 1;
-    priv->workAvailable = [[NSCondition alloc] init];
-    priv->suspendedCondition = [[NSCondition alloc] init];
-    priv->allWorkDone = [[NSCondition alloc] init];
-    priv->isSuspended = 0;
-
-    return self;
++ (NSOperationQueue*)mainQueue {
+    static StrongId<NSOperationQueue> mainQueue = [[NSOperationQueue_MainQueue new] autorelease];
+    return mainQueue;
 }
 
 /**
  @Status Interoperable
 */
-- (void)addOperation:(id)op {
-    /*
-    //  Add any dependencies
-    id dependencies = [op dependencies];
-    int count = [dependencies count];
-    for ( int i = 0; i < count; i ++ ) {
-    id curDep = [dependencies objectAtIndex:i];
-    [self addOperation:curDep];
+- (void)addOperation:(NSOperation*)op {
+    if (op.isExecuting || op.isFinished) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"operation is already or finished executing" userInfo:nil];
     }
-    */
-
-    [op _setCompletionQueue:_completionQueue];
 
-    unsigned priority = 1;
-    if ([op queuePriority] < NSOperationQueuePriorityNormal) {
-        priority = 2;
-    } else if ([op queuePriority] > NSOperationQueuePriorityNormal) {
-        priority = 0;
+    if (!op._acquirePermissionToAddToQueue) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"operation is already in an operation queue" userInfo:nil];
     }
 
-    NSAtomicListInsert(&priv->queues[priority], op);
-    [priv->workAvailable signal];
+    [self willChangeValueForKey:@"operations"];
+    [self willChangeValueForKey:@"operationCount"];
 
-    pthread_mutex_lock(&priv->_threadRunningLock);
-    if (priv->_thread == nil) {
-        priv->_thread = [[NSThread alloc] initWithTarget:self selector:@selector(_workThread) object:nil];
-        [priv->_thread start];
-        [priv->_thread release];
+    @synchronized(_operations.get()) {
+        [_operations addObject:op];
     }
-    pthread_mutex_unlock(&priv->_threadRunningLock);
-}
 
-/**
- @Status Interoperable
-*/
-- (void)addOperationWithBlock:(void (^)())block {
-    id op = [NSOperation new];
-    [op setCompletionBlock:block];
-    [self addOperation:op];
-    [op release];
+    [self didChangeValueForKey:@"operationCount"];
+    [self didChangeValueForKey:@"operations"];
+
+    [op addObserver:self forKeyPath:@"isFinished" options:0 context:nil];
```

Comment:
oh hm, i didn't think about this angle. i'll hold the lock on the operation then.